tests/framework/prototype.py

```diff
@@ -0,0 +1,279 @@
+                logger.info("--- RayClusterAddCREvent %s seconds ---" % (time.time() - start_time))
+                return
+            time.sleep(1)
+        raise Exception("RayClusterAddCREvent wait() timeout")
+
+class RayClusterDeleteCREvent(CREvent):
+    def exec(self):
+        client.CustomObjectsApi().delete_namespaced_custom_object(
+            group = 'ray.io', version = 'v1alpha1', namespace = self.namespace, plural = 'rayclusters', name = self.cr['metadata']['name'])
+
+    def wait(self):
+        start_time = time.time()
+        for _ in range(self.timeout):
+            headpods = client.CoreV1Api().list_namespaced_pod(namespace = self.namespace,
+                                                              label_selector='rayNodeType={}'.format('head'))
+            workerpods = client.CoreV1Api().list_namespaced_pod(namespace = self.namespace,
+                                                                label_selector='rayNodeType={}'.format('worker'))
+            if (len(headpods.items) == 0 and len(workerpods.items) == 0):
+                logger.info("--- RayClusterDeleteCREvent %s seconds ---" % (time.time() - start_time))
+                return
+            time.sleep(1)
+        raise Exception("RayClusterDeleteCREvent wait() timeout")
+
+# TestSuite
+class GeneralTestCase(unittest.TestCase):
+    def __init__(self, methodName, images, crEvent):
+        super(GeneralTestCase, self).__init__(methodName)
+        self.crEvent = crEvent
+        self.images = images
+
+    @classmethod
+    def setUpClass(cls):
+        delete_kind_cluster()
+
+    def setUp(self):
+        if not check_cluster_exist():
+            create_kind_cluster()        
+            install_crd()
+            download_images(self.images)
+            kind_load_images(self.images)
+            install_kuberay_operator()
+            config.load_kube_config()
+
+    def runTest(self):
+        self.crEvent.trigger()
+
+    def tearDown(self) -> None:
+        try:
+            delete_event = RayClusterDeleteCREvent(self.crEvent.cr, [], self.crEvent.timeout, self.crEvent.namespace)
+            delete_event.trigger()
+        except Exception as e:
+            logger.error(str(e))
+            delete_kind_cluster()
+
+if __name__ == '__main__':
+    template_name = 'config/ray-cluster.mini.yaml.template'
+    namespace = 'default'
+    with open(template_name) as base_yaml:
+        baseCR = yaml.load(base_yaml, Loader=yaml.FullLoader)
+
+    patch_list = [
+        jsonpatch.JsonPatch([{'op': 'replace', 'path': '/spec/headGroupSpec/template/spec/containers/0/name', 'value': 'ray-head-1'}]),
+        jsonpatch.JsonPatch([{'op': 'replace', 'path': '/spec/headGroupSpec/template/spec/containers/0/name', 'value': 'ray-head-2'}]),
+        # Reproduce #612
+        jsonpatch.JsonPatch([{'op': 'replace', 'path': '/spec/headGroupSpec/replicas', 'value': 2}]),
```

Comment:
For this one, are we validating that `headGrouSpec/replicas` is ignored?

Reply:
Yes, this test will fail in `RayClusterAddCREvent.wait()`. In the `wait` function, `expected_head_pods` is 2, but only 1 head pod exists in the Kubernetes cluster.

Reply:
Should we remove the test then?
(It's great though that you were able to use the test to re-discover our weirdness around head replicas.)

Reply:
Maybe we can keep this JsonPatch in this PR, and remove it when we integrate this framework into CI/CD? I mentioned #612 several times in the [design doc](https://docs.google.com/document/d/1OyMBgQ9zx5FF0GV3f3WKVdW-Y4OXz_epaywYgTw893w/edit?usp=sharing). If we remove the JsonPatch in this PR, it may cause the readers to be confused.

Reply:
Removed #612 from the design doc and the JsonPatch.